File: content/snippets/cpp/exercises/intermediate/word_frequency.cpp
#include <algorithm>
#include <cctype>
#include <iostream>
#include <sstream>
#include <string>
#include <unordered_map>
#include <vector>

int main() {
    std::string line = "This is a test. This test is simple.";
    for (char& c : line) {
        if (std::ispunct(static_cast<unsigned char>(c))) c = ' ';
        else c = static_cast<char>(std::tolower(static_cast<unsigned char>(c)));
    }

    std::unordered_map<std::string, int> freq;
    std::istringstream in(line);
    std::string word;
    while (in >> word) freq[word]++;

    std::vector<std::pair<std::string, int>> items(freq.begin(), freq.end());
    std::sort(items.begin(), items.end(), [](auto& a, auto& b) {
        return a.second == b.second ? a.first < b.first : a.second > b.second;
    });

    for (const auto& p : items) {
        std::cout << p.first << ": " << p.second << "\n";
    }
}

File: content/snippets/go/exercises/intermediate/word_frequency.go
package main

import (
	"fmt"
	"regexp"
	"sort"
	"strings"
)

func main() {
	text := "This is a test. This test is simple."
	re := regexp.MustCompile(`[A-Za-z]+`)
	words := re.FindAllString(strings.ToLower(text), -1)

	freq := map[string]int{}
	for _, w := range words {
		freq[w]++
	}

	type pair struct {
		word  string
		count int
	}

	items := make([]pair, 0, len(freq))
	for w, c := range freq {
		items = append(items, pair{w, c})
	}

	sort.Slice(items, func(i, j int) bool {
		if items[i].count == items[j].count {
			return items[i].word < items[j].word
		}
		return items[i].count > items[j].count
	})

	for _, it := range items {
		fmt.Printf("%s: %d\n", it.word, it.count)
	}
}
